File: Sort/Heap Sort/HeapSort.java
import java.util.Scanner;

public class HeapSort{
	public static void main(String[] args){
		Scanner s = new Scanner(System.in);
		System.out.print("Enter number of element: ");
		int array[] = new int[s.nextInt()];
		for(int i = 0; i < array.length; i++){
			System.out.print("Enter element at index " + i + " : ");
			array[i] = s.nextInt();
		}
		heapSort(array);
	}
	
	private static int[] heapSort(int[] array){
		for(int i = 0; i < array.length; i++){
			array = heapify(array, i);
		}
		
		for(int i = array.length - 1; i > 0; i--){
			array = deleteAndSort(array, i);
		}
		
		System.out.print("\n[");
		for(int i = 0; i < array.length; i++){
			System.out.print(" " + array[i] + " ");
		}
		System.out.println("]\n");
		
		return array;
	}
	
	private static int[] heapify(int[] array, int n){
		if(n != 0){
			if(array[n] > array[(n-1)/2]){
				int temp = array[n];
				array[n] = array[(n-1)/2];
				array[(n-1)/2] = temp;
				array = heapify(array, (n-1)/2);
			}
		}
		return array;
	}
	
	private static int[] deleteAndSort(int[] array, int n){
		int temp = array[n];
		array[n] = array[0];
		array[0] = temp;
		
		int i = 0;
		while(true){
			if((i + 1) * 2 < n){
				int maxIndex = (i + 1) * 2;
				if(array[maxIndex] < array[i * 2 + 1])
					maxIndex = i * 2 + 1;
				if(array[i] < array[maxIndex]){
					temp = array[maxIndex];
					array[maxIndex] = array[i];
					array[i] = temp;
				}
				i = maxIndex;
			}else{
				if(i * 2 + 1 < n){
					if(array[i] < array[i * 2 + 1]){
						temp = array[i * 2 + 1];
						array[i * 2 + 1] = array[i];
						array[i] = temp;
					}
				}
				break;
			}
		}
		return array;
	}
}

File: Sort/Heap Sort/HeapSort.kt
object HeapSort {
    fun sort(array: Array<Int>) {
        buildHeap(array)
        var i = array.size - 1
        var temp: Int
        while (i > 0) {
            temp = array[0]
            array[0] = array[i]
            array[i] = temp
            heapify(array, 0, i)
            i--
        }
    }

    private fun buildHeap(array: Array<Int>) {
        var i = (array.size - 1) / 2
        while (i >= 0) {
            heapify(array, i, array.size)
            i--
        }
    }

    private fun heapify(array: Array<Int>, n: Int, size: Int) {
        if (n * 2 + 1 < size) {
            if (array[n] < array[n * 2 + 1] || (n * 2 + 2 < size && array[n] < array[n * 2 + 2])) {
                val max = if (n * 2 + 2 < size && array[n * 2 + 1] <= array[n * 2 + 2]) {
                    n * 2 + 2
                } else {
                    n * 2 + 1
                }

                val temp = array[n]
                array[n] = array[max]
                array[max] = temp
                heapify(array, max, size)
            }
        }
    }
}